Spigot-API-Patches/0085-Add-extended-PaperServerListPingEvent.patch

```diff
@@ -0,0 +1,370 @@
+From 6eb251463799168d423fd32f96f32b56564bbe7a Mon Sep 17 00:00:00 2001
+From: Minecrell <minecrell@minecrell.net>
+Date: Wed, 11 Oct 2017 15:55:38 +0200
+Subject: [PATCH] Add extended PaperServerListPingEvent
+
+Add a new event that extends the original ServerListPingEvent
+and allows full control of the response sent to the client.
+
+diff --git a/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java b/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java
+new file mode 100644
+index 00000000..a6576f99
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java
+@@ -0,0 +1,319 @@
++package com.destroystokyo.paper.event.server;
++
++import static java.util.Objects.requireNonNull;
++
++import com.destroystokyo.paper.network.StatusClient;
++import com.destroystokyo.paper.profile.PlayerProfile;
++import org.bukkit.Bukkit;
++import org.bukkit.entity.Player;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.server.ServerListPingEvent;
++import org.bukkit.util.CachedServerIcon;
++
++import java.util.ArrayList;
++import java.util.Iterator;
++import java.util.List;
++import java.util.NoSuchElementException;
++import java.util.UUID;
++
++import javax.annotation.Nonnull;
++import javax.annotation.Nullable;
++
++/**
++ * Extended version of {@link ServerListPingEvent} that allows full control
++ * of the response sent to the client.
++ */
++public class PaperServerListPingEvent extends ServerListPingEvent implements Cancellable {
++
++    @Nonnull private final StatusClient client;
++
++    private int numPlayers;
++    private boolean hidePlayers;
++    @Nonnull private final List<PlayerProfile> playerSample = new ArrayList<>();
++
++    @Nonnull private String version;
++    private int protocolVersion;
++
++    @Nullable private CachedServerIcon favicon;
++
++    private boolean cancelled;
++
++    private boolean originalPlayerCount = true;
++    private Object[] players;
```

Comment:
Any reason this isn't just an array of Players?

It looks like it's only used for the Player iterator, so it's not like anyone will be putting anything else in it.
With a little massaging it'd also let us get rid of getBukkitPlayer and getOnlinePlayer below which seem oftly extraneous for this event.

Reply:
hmm I hadn't looked at the implementation when I wrote that. I'm not super keen on the style but I see why it exists.

Reply:
Basically I didn't want to make performance any worse than it was before. Originally, I had the `iterator()` method entirely in the implementation, but considering it's a bit tricky to get the implementation right in a backwards compatible way I figured having the base implementation in the API might be the best solution.

That way all implementations that support `PaperServerListPingEvent` implement the player iterator correctly, but still have the opportunity to optimize if needed (i.e. here: to avoid unnecessary wrapping).